picoquic/util.c

```diff
@@ -51,7 +52,7 @@ char* picoquic_string_create(const char* original, size_t len)
             }
             else if (allocated > len) {
                 memcpy(str, original, len);
-                str[len] = 0;
+                str[allocated - 1] = 0;
```

Comment:
If the intend is to set the character right after 'len' to '\0', I'd prefer the original version. The new version reads more like setting the last allocated character to '\0', which doesn't necessarily mean it's the character right after the copied string if you don't look at the whole function.

Reply:
That was the only way to silence the static analyzer of Visual Studio. I think there is something subtle going on with possible integer overflow.

Reply:
I think the Analyzer is just broken. If you try this obviously wrong code, it doesn't complain anymore:
```
        str = (char*)malloc(len);
        if (str != NULL) {
            memcpy(str, original, len);
            str[len] = 0;
        }
```

Reply:
I am a bit gun shy there. In Windows OS development, we used an internal version of the analyzer called Prefix -- similar to what ships in VS, but more capable. There were a few incidents were a bug was found "in the field", and postmortem inquiry showed that it was flagged by the static analyzer, but the developer discarded the warning as noise from a broken analyzer. Are you sure that the analyzer was actually running in your test? You need to have the "Control Flow Guard" option set to "Yes" in C/C++ "All options".

Reply:
Yes, "Control Flow Guard" option set to "Yes" in C/C++ "All options". I worked in util.c all the time and changed the code of `picoquic_string_create()`. You should be able to reproduce this by changing the code to the above version. Btw. I am running VS 2019 but with VS 2017 tool chain.